In [30]:
import pandas as pd
import numpy as np
import math
import datetime as dt
from datetime import timedelta, datetime

In [44]:
from dateutil.relativedelta import relativedelta

#### Получили файл набора данных, прилагающийся к этой работе dataset_1_Titanic.csv(60 КБ) на Kaggle

#### Считываем файл в pandas DataFrame:

In [115]:
titanic_df = pd.read_csv('train.csv')

In [83]:
titanic_df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

#### Количество точек данных (измерений) в наборе:

In [84]:
titanic_df.shape[0]

891

#### Количество полей данных (атрибутов):

In [85]:
titanic_df.shape[1]

12

#### Тип данных каждого поля, шкала каждого поля:

In [86]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


Таким образом, получаем следующую информацию:
<br>
- Поле **PassengerId**: тип **int64**, интервальная шкала
<br>
- Поле **Survived**: тип **int64**, номинальная шкала
<br>
- Поле **Pclass**: тип **int64**, номинальная шкала
<br>
- Поле **Name**: тип **ojbject**, номинальная шкала
<br>
- Поле **Sex**: тип **object**, номинальная шкала
<br>
- Поле **Age**: тип **float64**, номинальная шкала
<br>
- Поле **SibSp**: тип **int64**, номинальная шкала
<br>
- Поле **Parch**: тип **int64**, номинальная шкала
<br>
- Поле **Ticket**: тип **object**, номинальная шкала
<br>
- Поле **Fare**: тип **float64**, номинальная шкала
<br>
- Поле **Cabin**: тип **object**, номинальная шкала
<br>
- Поле **Embarked**: тип **object**, номинальная шкала

#### Количество отсутствующих значений для каждого поля:

In [87]:
titanic_df.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

- Поле **PassengerId**: 0 отсутствующих значений
<br>
- Поле **Survived**: 0 отсутствующих значений
<br>
- Поле **Pclass**: 0 отсутствующих значений
<br>
- Поле **Name**: 0 отсутствующих значений
<br>
- Поле **Sex**: 0 отсутствующих значений
<br>
- Поле **Age**: <font color='red'>177 отсутствующих значений</font>
<br>
- Поле **SibSp**: 0 отсутствующих значений
<br>
- Поле **Parch**: 0 отсутствующих значений
<br>
- Поле **Ticket**: 0 отсутствующих значений
<br>
- Поле **Fare**: 0 отсутствующих значений
<br>
- Поле **Cabin**: <font color='red'>687 отсутствующих значений</font>
<br>
- Поле **Embarked**: <font color='red'>2 отсутствующих значения</font>

#### Количество и названия константных и уникальных полей:

In [88]:
# уникальные поля
titanic_df.nunique()

PassengerId    891
Survived         2
Pclass           3
Name           891
Sex              2
Age             88
SibSp            7
Parch            7
Ticket         681
Fare           248
Cabin          147
Embarked         3
dtype: int64

In [89]:
# константные поля
titanic_df.Sex.unique()

array(['male', 'female'], dtype=object)

In [90]:
titanic_df.Pclass.unique()

array([3, 1, 2], dtype=int64)

In [91]:
titanic_df.Parch.unique()

array([0, 1, 2, 5, 3, 4, 6], dtype=int64)

In [92]:
titanic_df.Embarked.unique()

array(['S', 'C', 'Q', nan], dtype=object)

#### Удаляем из набора поля, не несущие информативной нагрузки - содержащие уникальные либо константные значения:

In [116]:
titanic_df.drop(['PassengerId','Name','Ticket','Cabin','Embarked'], axis = 1, inplace = True)

In [117]:
titanic_df.head()

Survived  Pclass     Sex   Age  SibSp  Parch     Fare
0         0       3    male  22.0      1      0   7.2500
1         1       1  female  38.0      1      0  71.2833
2         1       3  female  26.0      0      0   7.9250
3         1       1  female  35.0      1      0  53.1000
4         0       3    male  35.0      0      0   8.0500

#### Заменяем возраст пассажиров на год их рождения:

In [95]:
titanic_df['Age'].isna().sum()

177

In [118]:
titanic_df['Age'] = titanic_df['Age'].fillna(-1)
titanic_df['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
886    27.0
887    19.0
888    -1.0
889    26.0
890    32.0
Name: Age, Length: 891, dtype: float64

In [104]:
len(titanic_df['Age'])

891

In [119]:
dayX = dt.datetime(1912, 4, 15)
for i in range(len(titanic_df['Age'])):
    if (titanic_df['Age'][i] != -1):
        d = dayX - relativedelta(years = int(titanic_df['Age'][i]))
        titanic_df['Age'][i] = d.year

C:\Users\zstep\anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


In [120]:
titanic_df.head()

Survived  Pclass     Sex     Age  SibSp  Parch     Fare
0         0       3    male  1890.0      1      0   7.2500
1         1       1  female  1874.0      1      0  71.2833
2         1       3  female  1886.0      0      0   7.9250
3         1       1  female  1877.0      1      0  53.1000
4         0       3    male  1877.0      0      0   8.0500

#### Переименовываем все названия колонок на русском языке:

In [121]:
titanic_df.rename(columns = {'Survived': 'Выживание', 
                             'Pclass': 'Класс', 
                             'Sex': 'Пол', 
                             'Age': 'Год рождения', 
                             'SibSp': 'Братья, сестры или супруги', 
                             'Parch': 'Кол-во родителей/детей', 
                             'Fare': 'Размер тарифа'}, inplace = True)
titanic_df.head()

Выживание  Класс     Пол  Год рождения  Братья, сестры или супруги  \
0          0      3    male        1890.0                           1   
1          1      1  female        1874.0                           1   
2          1      3  female        1886.0                           0   
3          1      1  female        1877.0                           1   
4          0      3    male        1877.0                           0   

   Кол-во родителей/детей  Размер тарифа  
0                       0         7.2500  
1                       0        71.2833  
2                       0         7.9250  
3                       0        53.1000  
4                       0         8.0500

#### Заменяем все отсутствующие данные средним значением (в случае числового поля) либо самым распространенным в категориальных полях:

In [125]:
# в данном случае отсутствуют значения только в числовых полях, значит заменяем их средним
titanic_df.fillna(titanic_df.mean(), inplace = True)
titanic_df.isna().sum()

Выживание                     0
Класс                         0
Пол                           0
Год рождения                  0
Братья, сестры или супруги    0
Кол-во родителей/детей        0
Размер тарифа                 0
dtype: int64

In [126]:
# для категориальных использовалась бы команда:
# titanic_df.Embarked.fillna(titanic_df.Embarked.mode()[0], inplace = True)

#### Сохраняем получившийся набор данных в формате CSV:

In [263]:
titanic_df.to_csv('titanic_data.csv')

#### Открыли данный файл в программе Microsoft Excel и сохранили в формате .xlsx.

#### Считываем файл .xlsx в новый DataFrame:

In [279]:
titanic_new_df = pd.read_excel('titanic_data.xlsx', header = None, index_col = 0, sep = ',')

In [280]:
titanic_new_df.head()

Empty DataFrame
Columns: []
Index: [0,0,3,male,1890.0,1,0,7.25, 1,1,1,female,1874.0,1,0,71.2833, 2,1,3,female,1886.0,0,0,7.925, 3,1,1,female,1877.0,1,0,53.1, 4,0,3,male,1877.0,0,0,8.05]

#### Повторяем п. 2 для этого набора данных:

In [270]:
tit = pd.read_csv('titanic_data.csv', delimiter=',', header = 0, index_col = 0)

In [273]:
tit

Выживание  Класс     Пол  Год рождения  Братья, сестры или супруги  \
0            0      3    male        1890.0                           1   
1            1      1  female        1874.0                           1   
2            1      3  female        1886.0                           0   
3            1      1  female        1877.0                           1   
4            0      3    male        1877.0                           0   
..         ...    ...     ...           ...                         ...   
886          0      2    male        1885.0                           0   
887          1      1  female        1893.0                           0   
888          0      3  female          -1.0                           1   
889          1      1    male        1886.0                           0   
890          0      3    male        1880.0                           0   

     Кол-во родителей/детей  Размер тарифа  
0                         0         7.2500  
1                         0        71.2833  
2                         0         7.9250  
3                         0        53.1000  
4                         0         8.0500  
..                      ...            ...  
886                       0        13.0000  
887                       0        30.0000  
888                       2        23.4500  
889                       0        30.0000  
890                       0         7.7500  

[891 rows x 7 columns]

## Контрольные вопросы

### Какие две главные структуры данных используются в pandas? В чем их отличие?

**Series** и **DataFrame**:
<br>
**Series** - одномерная структура данных, т. е. объект-аналог списка или одномерного массива. 
<br>
**DataFrame** - табличная структура данных.

### Назовите основные функции чтения данных из файлов

1. Функция **read()** используется после функции **open()**. Ее аргумент - кол-во считываемых символов.
<br>
2. Функция **read_csv()** предназначена для чтения файла формата csv.
<br>
3. Функция **read_excel()** предназначена для чтения файла формата excel (.xls, .xlsx).
<br>
4. Функция **readline()** используется для считывания определенных строк.

### Напишите функцию записи набора данных А в файл с разделителями-запятыми

*A.to_csv('Путь на диске', sep = ',')*

### Как называется функция, выводящая на экран первые N строчек набора данных?

*A.head(N)*

### С помощью какой функции можно удалить поле набора данных? А строку?

Поле набор данных: 
<br>
*drop('Название поле набора данных', axis = 1)*
<br>
Строка: 
<br>
*drop('Название строки')*

### Напишите участок кода, демонстрирующий возможность добавления новой строки к набору данных. 

In [281]:
A = {'first column':[2, 4, 6], 'second column':[1, 3, 5], 'third column':['a', 'b', 'c']}
df = pd.DataFrame(A)
df

first column  second column third column
0             2              1            a
1             4              3            b
2             6              5            c

In [284]:
df.loc[df.shape[0]] = [8, 7, 'd']
df

first column  second column third column
0             2              1            a
1             4              3            b
2             6              5            c
3             5             10            c
4             8              7            d

### Напишите участок кода, демонстрирующий возможность добавления новой колонки (поля) к набору данных. 

In [285]:
A = {'first column':[2, 4, 6], 'second column':[1, 3, 5], 'third column':['a', 'b', 'c']}
df = pd.DataFrame(A)
df

first column  second column third column
0             2              1            a
1             4              3            b
2             6              5            c

In [286]:
df['fourth column'] = ['d', 'e', 'f']
df

first column  second column third column fourth column
0             2              1            a             d
1             4              3            b             e
2             6              5            c             f